### ISSUE TITLE
type.arguments.not.inferred skips structural checks, masking reference-checker qualifier diagnostics

### ISSUE THREAD

**Issue by wmdietl** [eisop/checker-framework#1980]
## Summary

In the jspecify reference-checker samples suite, a cluster of records (relayed count: 7) reports `type.arguments.not.inferred` where the sample expects either no diagnostic or a more specific one (e.g. an argument-vs-bound mismatch, `jspecify_nullness_mismatch`). Root cause: annotated type-argument inference legitimately fails under that checker's three-valued, non-reflexive-in-strict-mode qualifier lattice, and the caller then skips the invocation's structural checks entirely, so the specific qualifier diagnostic the sample wants never gets a chance to fire.

This issue is the CF-side mechanism analysis (independently verified against `master@1a360a6f6`, permalinks below), the concrete regression risk of the obvious fix, and a scoped recommendation — written up so a later effort doesn't have to re-derive any of it. **No fix is included here.** A related Javadoc inaccuracy this analysis surfaced is already fixed in #1979.

I do not have the reference-checker samples in the environment this was verified in, so the sample-level claims (the 7 records, their wanted diagnostics) are relayed, not independently checked here. Everything else below is independently verified by reading and by a reverted instrumentation trace.

### 1. A failed inference makes the caller skip all structural checks

Three call sites in `BaseTypeVisitor` call `checkTypeArgumentInference` and, on `false`, skip the invocation's structural (argument-vs-parameter and type-argument-vs-bound) checks entirely:

- `visitMethodInvocation`: https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java#L2104-L2108 (`return null`)
- `visitNewClass`: https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java#L2544-L2549 (`return null`)
- the member-reference path: https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java#L4611-L4615 (`return true`, i.e. treats the reference as conformant and stops)

`checkTypeArgumentInference` itself separates three failure modes — https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java#L2180-L2210 — `inferenceBudgetExceeded()` and `inferenceCrashed()` get their own error keys and `return false`; everything else falls to `reportTypeArgumentInferenceFailure`, the overridable hook the reference checker already consumes.

That hook's own doc comment already states the mechanism explicitly (worth noting: this isn't undocumented, it's just not the diagnostic being asked for): https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java#L2212-L2243

### 2. Why annotated inference fails when javac's does not

The engine is the JLS-18-style `typeinference8` package. `InferenceResult.inferenceFailed()` is set from `BoundSet.annoInferenceFailed`, turned on in exactly one place when a constraint reduces to the sentinel `ConstraintSet.TRUE_ANNO_FAIL`:

https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/framework/util/typeinference8/constraint/ConstraintSet.java#L448-L469

`TRUE_ANNO_FAIL` is a *TRUE* sentinel — unlike `FALSE`/`FalseBoundException`, it does not abort the solve; reduction and resolution continue, only the "annotations didn't work out" flag is raised. It's produced at two kinds of site, both encoding "the Java structure is fine, the qualifiers are not":

- `ProperType.isSubType` / `isSubTypeUnchecked` / `isAssignable` — the guard `types.isAssignable(...)` (precisely what javac checks) is true, but the checker's `getTypeHierarchy().isSubtype(subATM, superATM)` on the annotated types is false:
  - https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/framework/util/typeinference8/types/ProperType.java#L178-L198
  - https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/framework/util/typeinference8/types/ProperType.java#L212-L225
  - https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/framework/util/typeinference8/types/ProperType.java#L239-L252
- `QualifierTyping.reduceSubtyping` / `reduceEquality` — `isSubtypeQualifiersOnly(qAnno, rAnno)` is false for two concrete qualifiers. This is the site a non-reflexive lattice (`unspec` not `<:` `unspec` in strict mode) would trip; a standard reflexive lattice reaches the `ProperType.*` sites instead:
  - https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/framework/util/typeinference8/constraint/QualifierTyping.java#L85-L92
  - https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/framework/src/main/java/org/checkerframework/framework/util/typeinference8/constraint/QualifierTyping.java#L115-L125

So the mechanism is not "CF inference bails"; it's "the annotated subtype/qualifier relation embedded in a constraint is unsatisfiable, while the corresponding Java relation holds."

### 3. Live trace on a pure-CF, standard-lattice case (no reference checker needed)

A faithful pure-CF reproduction of the *problem* isn't constructible (it needs a non-reflexive qualifier order), but the *mechanism* and the *regression risk* both reproduce today on an ordinary reflexive lattice, via the existing test:

https://github.com/eisop/checker-framework/blob/1a360a6f60c0c0894c2a5fac07502ac1f651e6a3/checker/tests/tainting/SubtypingConstraint.java#L9-L15

```java
// :: error: (type.arguments.not.inferred)
asList(0, 0, "", Arrays.asList(o));
...
public static <T extends @Untainted Object> List<T> asList(T... a) { ... }
```

The `@Tainted` string literal forces `T = @Tainted`, violating the `@Untainted` bound. Temporarily instrumenting the two verification points above (prints at the `ProperType.isSubType` `TRUE_ANNO_FAIL` branch and at `checkTypeArgumentInference`'s failure return; not committed) and compiling that one file with the tainting checker produced:

```
[CFTRACE] ProperType.isSubType TRUE_ANNO_FAIL: Java assignable OK but annotated
          isSubtype(@Tainted Object, @Untainted Object) is false
... (repeated for @Tainted Integer <: @Untainted Object)
[CFTRACE] checkTypeArgumentInference returning FALSE (annoInferenceFailed) for:
          asList(0, 0, "", Arrays.asList(o)) -- caller will 'return null' and SKIP
          structural checks (checkTypeArguments/checkArguments).
          results-map-empty=false results-for-tree={T=@Tainted Object}
```

End-to-end confirmation: Java-level assignability succeeds, annotated subtyping fails, `annoInferenceFailed` is raised, and the caller skips the structural checks.

### 4. Best-effort type arguments are carried by the failed result

The trace's last line also settles a point the `InferenceResult` Javadoc got backwards before #1979: `results-map-empty=false`, `results-for-tree={T=@Tainted Object}`. Because `TRUE_ANNO_FAIL` doesn't abort the solve, `InvocationTypeInference.infer` still returns the instantiated variables, and resolution has genuinely instantiated `T` to the best-effort `@Tainted Object`. So best-effort arguments are already in hand at the failure point — confirmed, not hypothetical. (The Javadoc fix: #1979.)

## Assessment of the obvious fix ("fall through to structural checks")

The candidate fix: when annotated inference fails but javac's succeeds, instead of `return false`/skipping, run `checkTypeArguments`/`checkArguments` against the best-effort type arguments and let *those* checks decide the diagnostic.

**The precondition is real, not hypothetical:** best-effort arguments exist and are reachable (§4) — for the pure `annoInferenceFailed` case they're fully-resolved proper types, not half-solved variables. (This does **not** hold for the `inferenceCrashed`/`inferenceBudgetExceeded` paths, which pass `Collections.emptyList()` and must keep short-circuiting. Any fall-through must be gated to the pure `annoInferenceFailed` case.)

**But the trace also shows, concretely, why a blanket version regresses other tests.** In `SubtypingConstraint`, the best-effort `T = @Tainted Object` genuinely violates the `@Untainted` bound. If `checkTypeArguments` ran on it, the type-argument-vs-bound check would emit a bound/argument mismatch instead of `type.arguments.not.inferred` — and that test (and others like it — relayed count: ~23, from a prior blanket-downgrade attempt) *expects* `type.arguments.not.inferred`; converting the key is a regression by the samples' own acceptance criteria. The standard-lattice failures and the reference-checker failures arrive through the *identical* `annoInferenceFailed` path, and the best-effort argument is bound-violating in both — so there's no free separation between the two populations from the CF source alone.

**The crux, unresolved:** for the wanted-fix population, the wanted diagnostic *is* an argument/bound mismatch; for (at least some of) the ~23, the wanted answer is the generic `not.inferred`. Whether the structural check's output can be made to coincide with "wanted" in enough cases is empirical and needs the samples suite plus a prototype — not something the CF source alone can settle. Not prototyped (out of scope for the analysis pass, and the risk profile — a hot path in `BaseTypeVisitor` exercised by every generic invocation — warrants a dedicated effort).

## Recommendation, if picked up

Worth attempting only as a scoped, measured experiment, not a quick patch:

1. Gate strictly to `annoInferenceFailed` — never `inferenceCrashed`/`inferenceBudgetExceeded`, whose results carry no usable arguments.
2. Do it behind the existing `reportTypeArgumentInferenceFailure` seam / a checker opt-in, so default-checker behavior is untouched until proven.
3. Validation bar: the full samples suite in both modes **plus** the full `./gradlew alltests` (or at minimum `:checker:test` across the affected checkers) — because the mechanism reproduces on standard lattices, the change can silently re-key existing framework/checker tests (`checker/tests/tainting/SubtypingConstraint.java`, `Issue1111.java`, `checker/tests/regex/InvariantTypes.java`, and other `not.inferred` expectations across `checker/tests/`). Diff for any `not.inferred` expectation that flips to a different key — every such flip is a candidate regression to adjudicate, not a mechanical test-file update.

## Verified vs. relayed vs. open

- **Verified** (read + live trace, permalinks above, all pinned to `1a360a6f6`): the skip-on-failure control flow at all three call sites; `TRUE_ANNO_FAIL` as the sole source of `annoInferenceFailed` and its "Java-OK / annotations-not-OK" meaning; that best-effort args survive in the failed `InferenceResult`; that the mechanism and the regression risk both reproduce on a standard reflexive lattice via `SubtypingConstraint`.
- **Relayed** (not independently checked): the seven jspecify records and their wanted diagnostics; the "net −23" count from the prior blanket-downgrade attempt; the reference checker's non-reflexive strict-mode semantics.
- **Open:** whether a bound check on best-effort args can be made to coincide with the wanted diagnostic for the seven *without* re-keying the twenty-three. This is the crux and is unresolved — it needs the samples suite plus a prototype.

## Reproducing the trace

1. In `ProperType.isSubType`, print at the `else { return TRUE_ANNO_FAIL; }` branch (the `types.isAssignable` guard being true is javac's check succeeding).
2. In `BaseTypeVisitor.checkTypeArgumentInference`, print just before `reportTypeArgumentInferenceFailure`, including `args.getTypeArgumentsForExpression(tree)`.
3. `JAVA_HOME=<jdk21> ./gradlew assembleForJavac`, then `checker/bin/javac -processor org.checkerframework.checker.tainting.TaintingChecker checker/tests/tainting/SubtypingConstraint.java`.
4. Revert the prints.
